### DISCUSSION TITLE
Recommended way to use ANY?

### DISCUSSION THREAD

**Discussion by mmkal** [vitaly-t/pg-promise#907] [Q&A] [answered]
This works - is it safe?

```ts
await db.query("SELECT * FROM mytable WHERE id = ANY('{$1:csv}')", [1, 2, 3])
```

Or is there another recommended way of getting `SELECT * FROM mytable WHERE id = ANY('{1,2,3}')`?

The docs give an example of how to use `IN` with the csv filter, but I am migrating some existing code that uses `ANY` and I want to be sure I can get to parity before considering changing queries. Also `ANY` is more versatile, so I think there's a real use case.

**Comment by vitaly-t** [comment]
Is your question about how to format parameters differently? Please clarify what's the issue with the current code. This library is agnostic of the SQL syntax, as in it doesn't care if you use `IN` or `ANY` there. There is no such thing as recommended for your case specifically.

**Reply by mmkal** [reply]
Thank you for the response (and the library!)!

Sorry, I should have clarified, I meant how to make sure it is safe as far as SQL injection is concerned. Say, instead of hardcoded numbers `[1, 2, 3]` it was a list of strings that came straight from an untrusted client:

```ts
app.use((req, res) => {
  const result = await db.query("SELECT * FROM mytable WHERE id = ANY('{$1:csv}')", req.body.ids)
  res.send(result)
})
```

Having tried it, I see that it's not safe:

```ts
const req = { body: { ids: ["normal string", "'string with quotes'"] } }
await db.query("SELECT * FROM mytable WHERE id = ANY('{$1:csv}')", req.body.ids)
```

This results in a query with a syntax error, and doesn't even seem to include the second string which is surprising to me:

```sql
SELECT * FROM mytable WHERE id = ANY('{'normal string'}')
```

So this method I suggested, while it works for number inputs, doesn't work at all for strings. So even if there's no recommended way, could you point me to a way to use `ANY` with a list of string that works?

---

Edit: ah, my mistake, from experimenting it looks like I'm overthinking it, the ids just need to be passed as a single parameter:

```ts
const req = { body: { ids: ["normal string", "'string with quotes'"] } }
await db.query("SELECT * FROM mytable WHERE id = ANY($1)", [req.body.ids])
```

**Comment by vitaly-t** [comment, marked as answer]
In your code, you are making the same mistake as [here](https://stackoverflow.com/questions/58490759/filter-csv-from-pg-promise-takes-only-the-first-value-from-array) and [here](https://stackoverflow.com/questions/36839753/pass-an-array-of-integers-in-array-of-parameters/36847936#36847936).

Proper usage for CSV filter:

```ts
await db.query("SELECT * FROM mytable WHERE id = ANY($1:csv)", [req.body.ids])
```

or 

```ts
await db.query("SELECT * FROM mytable WHERE id = ANY(${ids:csv})", req.body)
```

or

```ts
await db.query("SELECT * FROM mytable WHERE id = ANY(${body.ids:csv})", req)
```

But if you want an `ARRAY` syntax there, then simply omit the `:csv` part.

**Reply by mmkal** [reply]
Thank you!
